File: src/main/java/dev/esophose/discordbot/listener/AutoRoleListener.java
package dev.esophose.discordbot.listener;

import dev.esophose.discordbot.Sparky;
import dev.esophose.discordbot.manager.GuildSettingsManager;
import discord4j.core.event.domain.Event;
import discord4j.core.event.domain.guild.MemberJoinEvent;
import discord4j.core.event.domain.role.RoleDeleteEvent;
import discord4j.core.object.entity.Member;
import discord4j.core.object.util.Snowflake;
import java.util.List;
import reactor.core.publisher.Mono;

public class AutoRoleListener extends Listener<Event> {

    private GuildSettingsManager guildSettingsManager;

    public AutoRoleListener() {
        super(MemberJoinEvent.class, RoleDeleteEvent.class);
    }

    @Override
    public void execute(Event event) {
        if (this.guildSettingsManager == null)
            this.guildSettingsManager = Sparky.getInstance().getManager(GuildSettingsManager.class);

        if (event instanceof MemberJoinEvent) {
            this.executeMemberJoin((MemberJoinEvent) event);
        } else if (event instanceof RoleDeleteEvent) {
            this.executeRoleDelete((RoleDeleteEvent) event);
        }
    }

    private void executeMemberJoin(MemberJoinEvent event) {
        if (event.getMember().isBot())
            return;

        List<Snowflake> autoRoleIds = this.guildSettingsManager.getGuildSettings(event.getGuildId()).getAutoRoleIds();
        if (!autoRoleIds.isEmpty()) {
            Member member = event.getMember();
            Mono<Void> mono = null;
            for (Snowflake roleId : autoRoleIds) {
                if (mono == null) {
                    mono = member.addRole(roleId);
                } else {
                    mono = mono.then(member.addRole(roleId));
                }
            }
            if (mono != null)
                mono.subscribe();
        }
    }

    private void executeRoleDelete(RoleDeleteEvent event) {
        List<Snowflake> autoRoleIds = this.guildSettingsManager.getGuildSettings(event.getGuildId()).getAutoRoleIds();
        if (autoRoleIds.contains(event.getRoleId()))
            this.guildSettingsManager.removeAutoRole(event.getGuildId(), event.getRoleId());
    }

}

File: src/main/kotlin/dev/esophose/discordbot/listener/AutoRoleListener.kt
package dev.esophose.discordbot.listener

import dev.esophose.discordbot.Sparky
import dev.esophose.discordbot.manager.GuildSettingsManager
import discord4j.common.util.Snowflake
import discord4j.core.`object`.entity.Member
import discord4j.core.event.domain.Event
import discord4j.core.event.domain.guild.MemberJoinEvent
import discord4j.core.event.domain.guild.MemberUpdateEvent
import discord4j.core.event.domain.role.RoleDeleteEvent
import reactor.core.publisher.Mono

class AutoRoleListener : Listener<Event>(MemberJoinEvent::class, MemberUpdateEvent::class, RoleDeleteEvent::class) {

    private var guildSettingsManager: GuildSettingsManager? = null

    override fun execute(event: Event) {
        if (this.guildSettingsManager == null)
            this.guildSettingsManager = Sparky.getManager(GuildSettingsManager::class)

        when (event) {
            is MemberJoinEvent -> executeMemberJoin(event)
            is MemberUpdateEvent -> executeMemberUpdate(event)
            is RoleDeleteEvent -> executeRoleDelete(event)
        }
    }

    private fun grantRole(guildId: Snowflake, member: Member) {
        val autoRoleIds = this.guildSettingsManager!!.getGuildSettings(guildId).autoRoleIds
        if (autoRoleIds.isNotEmpty()) {
            var mono: Mono<Void>? = null
            for (roleId in autoRoleIds) {
                mono = if (mono == null) {
                    member.addRole(roleId)
                } else {
                    mono.then(member.addRole(roleId))
                }
            }
            mono?.subscribe()
        }
    }

    private fun executeMemberJoin(event: MemberJoinEvent) {
        if (!event.member.isBot && !event.member.isPending)
            this.grantRole(event.guildId, event.member)
    }

    private fun executeMemberUpdate(event: MemberUpdateEvent) {
        if (event.old.isEmpty || !event.old.get().isPending)
            return

        event.member.subscribe { member ->
            if (!member.isPending)
                this.grantRole(event.guildId, member)
        }
    }

    private fun executeRoleDelete(event: RoleDeleteEvent) {
        val autoRoleIds = this.guildSettingsManager!!.getGuildSettings(event.guildId).autoRoleIds
        if (autoRoleIds.contains(event.roleId))
            this.guildSettingsManager!!.removeAutoRole(event.guildId, event.roleId)
    }

}
